jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -0,0 +1,115 @@
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+
+/** Class used for inferring the main class in an application. */
+public class MainClassFinder {
+
+  /** Helper class for loading a .class file. */
+  private static class ClassFileLoader extends ClassLoader {
+
+    private Path classFile;
+
+    private ClassFileLoader(Path classFile) {
+      this.classFile = classFile;
+    }
+
+    @Nullable
+    @Override
+    public Class findClass(String name) {
+      try {
+        byte[] bytes = Files.readAllBytes(classFile);
+        return defineClass(name, bytes, 0, bytes.length);
+      } catch (IOException | ClassFormatError ignored) {
+        // Not a valid class file
+        return null;
+      }
+    }
+  }
+
+  /**
+   * Searches for a class containing a main method given an absolute root directory.
+   *
+   * @return the name of the class if one is found, null if no class is found.
+   * @throws IOException if searching/reading files fails.
+   * @throws MultipleClassesFoundException if more than one valid main class is found.
+   */
+  @Nullable
+  public static String findMainClass(String rootDirectory)
+      throws MultipleClassesFoundException, IOException {
+    // Make sure rootDirectory is valid
+    if (!Files.exists(Paths.get(rootDirectory)) || !Files.isDirectory(Paths.get(rootDirectory))) {
+      return null;
+    }
+
+    String className = null;
+    try (Stream<Path> pathStream = Files.walk(Paths.get(rootDirectory))) {
+      // Get all .class files
+      List<Path> classFiles =
+          pathStream
+              .filter(Files::isRegularFile)
+              .filter(path -> path.toString().endsWith(".class"))
+              .collect(Collectors.toList());
+
+      for (Path classFile : classFiles) {
+        // Convert filename (rootDir/path/to/ClassName.class) to class name (path.to.ClassName)
+        String name = classFile.toAbsolutePath().toString();
+        name = name.substring(rootDirectory.length() + 1);
+        name = name.replace('/', '.').replace('\\', '.');
+        name = name.substring(0, name.length() - ".class".length());
+
+        Class fileClass = new ClassFileLoader(classFile).findClass(name);
+        if (fileClass == null) {
+          // Got an invalid class file, keep searching
+          continue;
+        }
+
+        try {
+          // Check if class contains a public static void main(String[] args)
+          Method main = fileClass.getMethod("main", String[].class);
+          if (main != null
+              && main.getReturnType() == void.class
+              && Modifier.isStatic(main.getModifiers())
+              && Modifier.isPublic(main.getModifiers())) {
+            if (className == null) {
+              // Main method found; save it and continue searching for duplicates
+              className = name;
+            } else {
+              // Found more than one main method
+              throw new MultipleClassesFoundException(className, name);
```

Comment:
An alternative could be to change this method to return a list of main classes found so that

1. No need to have this exception
2. Doesn't need to be `@Nullable`.